filters/auth/oauth.go

```diff
@@ -0,0 +1,343 @@
+	}
+
+	ts := make([]string, len(t))
+	for i, ti := range t {
+		ts[i] = ti.Id
+	}
+
+	return ts, nil
+}
+
+func newSpec(typ roleCheckType) filters.Spec {
+	return &spec{typ: typ}
+}
+
+// Creates a new auth filter specification to validate authorization
+// tokens, optionally check realms and optionally check scopes.
+//
+// authUrlBase: the url of the token validation service.
+// The filter expects the service to validate the token found in the
+// Authorization header and in case of a valid token, it expects it
+// to return the user id and the realm of the user associated with
+// the token ('uid' and 'realm' fields in the returned json document).
+// The token is set as the Authorization Bearer header.
+//
+func NewAuth() filters.Spec {
+	return newSpec(checkScope)
+}
+
+// Creates a new auth filter specification to validate authorization
+// tokens, optionally check realms and optionally check groups.
+//
+// authUrlBase: the url of the token validation service. The filter
+// expects the service to validate the token found in the Authorization
+// header and in case of a valid token, it expects it to return the
+// user id and the realm of the user associated with the token ('uid'
+// and 'realm' fields in the returned json document). The token is set
+// as the Authorization Bearer header.
+//
+// groupUrlBase: this service is queried for the group ids, that the
+// user is a member of ('id' field of the returned json document's
+// items). The user id of the user is appended at the end of the url.
+//
+func NewAuthGroup() filters.Spec {
+	return newSpec(checkGroup)
+}
+
+func (s *spec) Name() string {
+	if s.typ == checkScope {
+		return AuthName
+	} else {
+		return AuthGroupName
+	}
+}
+
+func (s *spec) CreateFilter(args []interface{}) (filters.Filter, error) {
+	sargs, err := getStrings(args)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(sargs) == 0 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	var ac *authClient
+	ac, sargs = &authClient{sargs[0]}, sargs[1:]
+
+	var tc *groupClient
+	if s.typ == checkGroup {
+		tc, sargs = &groupClient{sargs[0]}, sargs[1:]
+	}
+
+	f := &filter{typ: s.typ, authClient: ac, groupClient: tc}
+	if len(sargs) > 0 {
+		f.realm, f.args = sargs[0], sargs[1:]
+	}
+
+	return f, nil
+
+}
+
+func (f *filter) validateRealm(a *authDoc) bool {
+	if f.realm == "" {
+		return true
+	}
+
+	return a.Realm == f.realm
+}
+
+func (f *filter) validateScope(a *authDoc) bool {
+	if len(f.args) == 0 {
+		return true
+	}
+
+	return intersect(f.args, a.Scopes)
+}
+
+func (f *filter) validateGroup(token string, a *authDoc) (bool, error) {
+	if len(f.args) == 0 {
+		return true, nil
+	}
+
+	groups, err := f.groupClient.getGroups(a.Uid, token)
+	return intersect(f.args, groups), err
+}
+
+func (f *filter) Request(ctx filters.FilterContext) {
+	r := ctx.Request()
+
+	token, err := getToken(r)
+	if err != nil {
+		unauthorized(ctx, "", missingBearerToken)
+		return
+	}
+
+	a, err := f.authClient.validate(token)
```

Comment:
"a" is an "authDoc" right?
I would like to have more clear variable name here, because it's not an index variable and it's derived from a call to a method, where the sicgnature is not easy to understand from this point in code.